libmisc/copydir.c

```diff
@@ -47,40 +47,43 @@ struct link_name {
 };
 static /*@exposed@*/struct link_name *links;
 
-static int copy_entry (const char *src, const char *dst,
+struct path_info {
+	const char *full_path;
+	int dirfd;
+	const char *name;
```

Comment:
Is it safe to assume that name should be the relative path of full_path under dirfd?

If that's the case, then copy_tree() using the same path for full_path and name, while AT_FDCWD for dirfd, seems wrong, unless we've guarnteed to chdir("/") first.

Reply:
Yeah, sorry, shadow's use of globals here kills me.  I'll reread this in a bit.

Reply:
But FWIW it still does look wrong to me.  For instance, copy_tree_impl() does

```
if (   fstatat (dst->dirfd, dst->name, &sb, 0) == 0
```

but 

```
957 int copy_tree (const char *src_root, const char *dst_root,
958                bool copy_root, bool reset_selinux,
959                uid_t old_uid, uid_t new_uid,
960                gid_t old_gid, gid_t new_gid)
961 {
962         const struct path_info src = {
963                 .full_path = src_root,
964                 .dirfd = AT_FDCWD,
965                 .name = src_root
966         };
```

So if cwd is, say /home, then the open will fail IIUC.

Reply:
I am not sure I understand your question.
This particular `fstatat()` call should fail as the destination should not already exist, as in https://github.com/shadow-maint/shadow/blob/4b3dde0b11314266c92e33cd9e95c28d4ea5d586/libmisc/copydir.c#L222-L224

Reply:
My question is about the AT_FDCWD.  Your copy_entry() will do:

```
423         if (fstatat(src->dirfd, src->name, &sb, AT_SYMLINK_NOFOLLOW) == -1) {
424                 /* If we cannot stat the file, do not care. */
```

So if it cannot open src->name, as a relative path under src->dirfd, then it will not copy the file.  So if src->dirfd is not the root to be copied from, then the fstatat will fail and the file will not be copied.

Even before that, copy_tree() -> copy_tree_impl() -> if (copy_root) will attempt:

```
263                 if (fstatat (src->dirfd, src->name, &sb, AT_SYMLINK_NOFOLLOW) =
    = -1) {
264                         return -1;
265                 }
```

dirfd is always set to AT_FDCWD and name is set to src_root, the full path, but there is no chdir in src/useradd.c or libmisc/copydir.

So if you "cd /etc; useradd ...", it appears to me that the copyfile will fail.

Reply:
Then again, I see that the manpages say:

       If pathname is absolute, then dirfd is ignored.